File: Amazon Spring '23 High Frequency/Optimal Partition of String/solution.cpp
// First Approach

class Solution
{
public:
    int partitionString(string s)
    {
        set<char> st;
        int count = 1;
        for (auto c : s)
        {
            auto pos = st.find(c);
            if (pos != st.end())
            {
                count++;
                st.clear();
            }
            st.insert(c);
        }
        return count;
    }
};

// Best Approach
class Solution
{
public:
    int partitionString(string s)
    {
        vector<bool> flag(26, false);
        int i = 0, count = 1;
        while (i < s.length())
        {
            int n = s[i] - 'a';
            if (flag[n] == true)
            {
                count++;
                fill(flag.begin(), flag.end(), false);
            }
            flag[n] = true;
            i++;
        }
        return count;
    }
};

File: Amazon Spring '23 High Frequency/Optimal Partition of String/solution.py
class Solution:
    def partitionString(self, s: str) -> int:

        hash = [False] * 26

        i = 0
        ans = 1
        while i < len(s):
            n = ord(s[i]) - ord('a')
            if hash[n] == True:
                hash = [False] * 26
                ans += 1
            hash[n] = True
            i += 1
        return ans

        